juju/juju
# Remove dead spaces

## Description of change

Attempt to remove dead spaces when performing reload spaces. At the
moment it only attempts to remove the spaces if they're dead. That way
we can ensure that we don't remove something accidentally.

The change is somewhat radical, in that it ensures the death
of a space if it's alive. That way we get what we want by removing the
space.

This isn't the way we want it to work in the future but for now this
the solution works the way we think it makes sense right now.

Future solutions should take more care around the topology and the
connections between nodes in the graph.

----

Additionally, we also need to carefully look at the model endpoint
bindings per application when we're attempting to remove a space. As
spaces might be in use else were and because we don't have the Life
checks for spaces in the code, we can't guarantee we won't break
something there.

There might be other places we also need to check, but the current 
rundown is:

 1. Don't delete if you're the alpha space
 2. Don't delete if you're the default space (model config)
 3. Don't delete if you're in a model endpoint binding
 4. Don't delete if you're a constraint of a model

Hopefully, these should cover us for now, but we should be more ruthless
when deleting if spaces were to check the life before use.

## QA steps

Tested that we didn't regress AWS, as that doesn't know about spaces, only
subnets. We're good there.

We can follow the same core steps though.

In MAAS create a new space (VLAN in MAAS so we can delete it)

```sh
juju spaces
juju reload-spaces
juju spaces # See that it's showed up
juju add-space isolated XX.XX.XX.XX/XX # Add the space
```

Delete the space from within MAAS

```sh
juju reload-spaces
```

## Documentation changes

The deletion of spaces in MAAS now reflects in Juju after reload-spaces.
